tests: POSIX: Refactor POSIX API tests

spoorthik opened the pull request:
Refactor and club the POSIX API tests into one test suite.
Enhance tests to improve code coverage.

Signed-off-by: Spoorthi K <spoorthi.k@intel.com>

nniranjhana commented on tests/posix/semaphore/prj.conf:
```diff
@@ -1,4 +0,0 @@
-CONFIG_PTHREAD_IPC=y
-CONFIG_MAIN_THREAD_PRIORITY=2
```
This `CONFIG_MAIN_THREAD_PRIORITY` from semaphore test is not included in the new unified prj.conf.. is it not necessary? Does the sema test behavior change in any way without it?

spoorthik replied:
@nniranjhana : With unified test suite, changing main thread priority would be misleading and also there is no need of changing the thread priority even for semaphores. The behavior is same for all tests with and without main thread priority 2.

spoorthik commented on tests/posix/src/semaphore.c:
```diff
@@ -8,22 +8,20 @@
-static void *foo_func(void *p1)
+static void *child_func(void *p1)
 {
-	printk("Child thread running\n");
```
@nniranjhana: Removed as this statement didn't provide more information on result of execution of semaphore tests as such.

spoorthik commented:
recheck

spoorthik committed:
tests: POSIX: Refactor POSIX API tests

Refactor and club the POSIX API tests into one test suite.

Signed-off-by: Spoorthi K <spoorthi.k@intel.com>

spoorthik committed:
tests: POSIX: Enhance semaphore tests

Enhance semaphore test to validate sem_timedwait()
    and sem_trywait(). This is intended to improve
    code coverage.

Also modify ztest check with zassert_equal with
return value instead of zassert_false for better
understandability.

Signed-off-by: Spoorthi K <spoorthi.k@intel.com>

nniranjhana commented on tests/posix/src/semaphore.c:
```diff
@@ -28,32 +28,46 @@ void test_posix_semaphore(void)
+	zassert_equal(ret, 0, "Thread creation failed");
+
+	zassert_equal(sem_timedwait(&sema, &abstime), -1, NULL);
+	zassert_equal(errno, ETIMEDOUT, NULL);
```
Add `sem_timedwait` with a successful return too, same for `sem_trywait` below

spoorthik replied:
For successful return, I would be required to change the current thread priority to preemptive. Is there any API in POSIX which changes the current thread priority?

nniranjhana replied:
I'm not sure why you'd need to change the current thread priority here, but sure you can do it by calling `pthread_setschedparam` with the pthread_self after setting `schedparam.priority` appropriately.

Without that, you can move the call to `sched_yield` above, so the child thread does the `sem_post` and makes timed_wait successful. We can then play around with giving and taking multiple times like in the CMSIS semaphore test.

spoorthik replied:
Well, the reason why I asked function to change priority was, the timedwait was timing out without pending the thread. Later figured out that the issue was with the usage of timeout value specified in sem_timedwait. Anyways, your intention of code coverage is addressed now. Please have a look. Thanks!

nashif commented on tests/posix/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_PTHREAD_IPC=y
```
that collection of tests should be in a subdirectory as well,, we should not have it at this level, call it tests/posix/common or something like that.

nashif replied:
are you planning to move everything else into this one single test?

nniranjhana commented on tests/posix/common/src/semaphore.c:
```diff
@@ -0,0 +1,95 @@
+	zassert_equal(sem_trywait(&sema), -1, NULL);
+	zassert_equal(errno, EAGAIN, NULL);
+
+	zassert_equal(sem_destroy(&sema), 0, "semaphore is not destroyed");
```
You could replace lines 77 to 87 with a `sem_post(&sema)`: I mean to say, reuse both semaphore and pthread attributes without having to delete and initialize again - it has the same effect. You can observe this in your initializing thread attributes: you haven't set the `schedparam.priority` but it still works in creating the thread with same previous priority, as the thread resources are still available, and are only released with `pthread_join` or `pthread_detach`.

spoorthik replied:
But the child thread might have terminated already by the time we call sem_post. And trying to have synchronization without having threads to synchronize would logically may not make sense. Also schedparam.priority has nothing to do with pthread_attr_destroy. It is just a structure with one of its member's initialized.

nniranjhana replied:
Okay, agreed on the logically having two threads for synchronizing - but can we have the same semaphore for doing it? Can `sched_yield`, make the child thread give the semaphore. Just check if deleting and re-initializing is really needed.

spoorthik replied:
As said earlier, sched_yield() will not make any difference when the child thread has already terminated. To avoid confusion, I have updated the code with comments. Also since I am checking all success and failure conditions, I would require 2 threads and hence initializing the attributes. Please have a look at latest patch. Thanks.

nniranjhana commented on tests/posix/common/src/semaphore.c:
```diff
@@ -0,0 +1,95 @@
+
+static void *child_func(void *p1)
+{
+	zassert_equal(sem_post(dummy_sem), -1, "sem_post of"
```
Good call on the dummy semaphore check, however you can move this to the start of the `test_posix_semaphore` function; else we test this twice when calling the child thread again. We only need the successful `sem_post`, don't we?

spoorthik replied:
Why can't we call the same thread entry function while creating 2 threads? Anyways it not adding any overhead in terms of execution time right?

nniranjhana replied:
I was not talking about calling the thread entry function twice. I was talking about the code from lines 21 to 23 being executed twice.

nniranjhana reviewed (commented):
Looks good, coverage has improved a lot as you've covered almost all the negative scenarios - just a couple of optimization reviews to be addressed.

We also have two cases in `sem_destroy` (i) add the check for the same `dummy_sem` you've added to return EINVAL and (ii) just call it once more after semaphore is destroyed, it should cover the EBUSY case.

spoorthik commented:
@nashif, @nniranjhana, @ramakrishnapallala : Please review the latest patch.
